include/extractor/guidance/mergable_road_detector.hpp

```diff
@@ -0,0 +1,149 @@
+    using MergableRoadData = IntersectionShapeData;
+
+    MergableRoadDetector(const util::NodeBasedDynamicGraph &node_based_graph,
+                         const std::vector<QueryNode> &node_coordinates,
+                         const IntersectionGenerator &intersection_generator,
+                         const CoordinateExtractor &coordinate_extractor,
+                         const util::NameTable &name_table,
+                         const SuffixTable &street_name_suffix_table);
+
+    /*
+     * OSM ways tend to be modelled as separate ways for different directions. This is often due to
+     * small gras strips in the middle between the two directions or due to pedestrian islands at
+     * intersections.
+     *
+     * To reduce unnecessary information due to these artificial intersections (which are not
+     * actually perceived as such) we try and merge these for our internal representation to both
+     * get better perceived turn angles and get a better reprsentation of our intersections in
+     * general.
+     *
+     *         i   h                                             i,h
+     *         |   |                                              |
+     *         |   |                                              |
+     * b - - -   v   - - - g                                      |
+     *         > a <               is transformed into: b,c - - - a - - -  g,f
+     * c - - -   ^   - - - f                                      |
+     *         |   |                                              |
+     *         |   |                                              |
+     *         d   e                                             d,e
+     */
+    bool CanMergeRoad(const NodeID intersection_node,
+                      const MergableRoadData &lhs,
+                      const MergableRoadData &rhs) const;
+
+    /* check if a road cannot influence the merging of the other. This is necessary to prevent
+     * situations with more than two roads that could participate in a merge */
+    bool IsDistinctFrom(const MergableRoadData &lhs, const MergableRoadData &rhs) const;
+
+  private:
+    // check if two name ids can be seen as identical (in presence of refs/others)
+    bool NameIDsAreCompatible(const NameID lhs, const NameID rhs) const;
+
+    /* When it comes to merging roads, we need to find out if two ways actually represent the
+     * same road. This check tries to identify roads which are the same road in opposite directions
+     */
+    bool RoadDataIsCompatible(const util::NodeBasedEdgeData &lhs_edge_data,
+                              const util::NodeBasedEdgeData &rhs_edge_data) const;
+
+    /* Detect traffic loops.
+     * Since OSRM cannot handle loop edges, we cannot directly see a connection between a node and
+     * itself. We need to skip at least a single node in between. */
+    bool IsTrafficLoop(const NodeID intersection_node, const MergableRoadData &road) const;
```

Comment:
Hm. This could probably be used at some other location. I'd prefer to wait for other refactors when it comes up, though.